test/unittests/execute_floating_point_test.cpp

```diff
@@ -42,6 +42,134 @@ MATCHER_P(ArithmeticNaN, value, "result with an arithmetic NaN")
     return FP<value_type>{result_value}.is_arithmetic_nan();
 }
 
+
+static constexpr auto all_rounding_directions = {
+    FE_TONEAREST, FE_DOWNWARD, FE_UPWARD, FE_TOWARDZERO};
+
+template <typename T>
+class TestValues
+{
+    using Limits = typename FP<T>::Limits;
+
+    inline static const std::array m_values{
+        T{0.0},
+
+        Limits::denorm_min(),
+        std::nextafter(Limits::denorm_min(), Limits::infinity()),
+        std::nextafter(Limits::min(), T{0}),
+        Limits::min(),
+        std::nextafter(Limits::min(), Limits::infinity()),
+        std::nextafter(T{1.0}, T{0}),
+        T{1.0},
+        std::nextafter(T{1.0}, Limits::infinity()),
+        std::nextafter(Limits::max(), T{0}),
+        Limits::max(),
+
+        Limits::infinity(),
+
+        // Canonical NaN:
+        FP<T>::nan(FP<T>::canon),
+
+        // Arithmetic NaNs:
+        FP<T>::nan((FP<T>::canon << 1) - 1),             // All bits set.
+        FP<T>::nan(FP<T>::canon | (FP<T>::canon >> 1)),  // Two top bits set.
+        FP<T>::nan(FP<T>::canon + 1),
+
+        // Signaling (not arithmetic) NaNs:
+        FP<T>::nan(FP<T>::canon >> 1),  // "Standard" signaling NaN.
+        FP<T>::nan(2),
+        FP<T>::nan(1),
+    };
+
+public:
+    using Iterator = typename decltype(m_values)::const_iterator;
+
+    static constexpr auto num_nans = 7;
+    static constexpr auto num_positive = m_values.size() - num_nans;
```

Comment:
I would suggest to simplify by declaring constants like
```
static const Iterator first_non_zero = &m_values[1];
static const Iterator first_nan = &m_values[num_positive];
static const Iterator first_noncanonical_nan = &m_values[num_positive + 1];
static const Iterator infinity = &m_values[num_positive - 1];
```

Then `positive_nonzero_finite` is `{first_non_zero, infinity}`,
`positive_nonzero_infinit` is `{fist_non_zero, first_nan}`
etc.

Reply:
Done.